### PR TITLE
sft_aet_giuh_merge

### PR DESCRIPTION
**Pull request by ajkhattak**
This PR merges changes to cfe related to soil freeze-thaw coupling, giuh refactoring, and rootzone-based AET which couples soil moisture profile module to cfe. Multiple branches have been merged into a single branch, leading to this PR. The code has been tested on different machines with and without changes using pseudo and ngen frameworks. Runs with changes reproduce the old results when coupling flags are turned off. By default the coupling flags are OFF, so no changes are needed to the existing cfe config files.

**Note: To test this PR on ngen framework, please checkout ngen/[cfe-modifications](https://github.com/NOAA-OWP/ngen/tree/ajk/cfe-modifications) branch due to the inclusion of SLoTH bmi to the realization files.**

## Additions
- Modifications to runoff schemes (Schaake and Xinanjiang)
- Added giuh.c to take giuh part out of cfe
- Rootzone-based AET which modifies et_from_soil 
- Added options for turning ON cfe coupling to other others (SFT and SMP)
- New config files are added for the AET rootzone-based example
- cmake based build

## Removals
- The part computing giuh runoff has been put into a separate file giuh.c
- individual make_and_run files have been deleted

## Changes
- No fundamental changes to cfe if run in standalone mode
- Changes to runoff schemes due to ice_fraction inclusion (active only if coupled to SoilFreezeThaw)
- Changes to soil AET function to include rootzone-based AET (active only If coupled to SoilMoistureProfile)
- Added CMakeFile.txt to build and run cfe with different options (see README for more details)

## Testing
1. tested in the standalone mode 
2. tested in the pseudo framework
3. tested in the ngen frameowrk
4. tested on Linux and macOS machines

## Notes
- Due to the addition of new input parameters (such as ice_fraction_schaake) to the BMI, the config files for running in the ngen framework have been updated (a separate PR will be submitted to ngen repo; [ajk/cfe-modifications](https://github.com/NOAA-OWP/ngen/tree/ajk/cfe-modifications)). The SLoTH bmi is used to provide default (or zero) values to the CFE BMI input parameters that are not active (i.e., neither used by cfe nor provided by any other BMI)

## Todos
- Update README (minor change)

## Checklist

- [x] PR has an informative and human-readable title
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows project standards (link if applicable)
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [x] Windows
- [x] Linux
- [x] Mac

### CODE REVIEW

**Comment by ajkhattak** [comment]
@SnowHydrology @rlmcdaniel @madMatchstick @lcunha0118 
Please see the below comment, which I missed to add in the initial submission.
**Note: To test this PR on ngen framework, please checkout [cfe-modifications](https://github.com/NOAA-OWP/ngen/tree/ajk/cfe-modifications) ngen repo branch due to the inclusion of SLoTH bmi to the realization files. ** This ngen branch will be merged to the ngen master after we merge this PR here.

File: README.md [hunk 1]
```diff
@@ -1,18 +1,36 @@
...
 
-## Compiling and running CFE
+## Compiling and Running CFE
+There are four examples for running CFE as described below. They assume you have [GCC](https://gcc.gnu.org) and [CMAKE](https://cmake.org/) on your machine. To build without cmake see section `Alternative: Compiling and Running CFE` below.
+
+1. `Option BASE` : CFE reads its own forcing file (standalone CFE BMI run; one BMI example)
+2. `Option FORCING` : CFE uses an external module to read in a forcing file and pass those data using BMI (two BMIs example)
+3. `Option FORCINGPET` : CFE uses external modules to read in a forcing file and calculate potential evapotranspiration, and pass those data using BMIs (three BMIs example)
+4. `Option AETROOTZONE` : Option #3 running with rootzone-based actural evapotranspiration (AET). This version of CFE is coupled to the [SoilMoistureProfiles](https://github.com/NOAA-OWP/SoilMoistureProfiles) module which provides soil moisture (1D array) that is used to estimate actual evapotranspiration (AET) from the deepest rootzone layer. This example requires the user to clone [SoilMoistureProfiles](https://github.com/NOAA-OWP/SoilMoistureProfiles) repo. (four BMIs example)
```
Comment (rlmcdaniel) [inline README.md:17]: "actual" instead of "actural" evapotranspiration

File: README.md [hunk 2]
```diff
@@ -44,79 +62,75 @@ A [configs/](./configs/) directory contains primiary configuration text files fo
 | a_Xinanjiang_inflection_point_parameter | *double* |   |  | parameter_adjustable | direct runoff | when `surface_partitioning_scheme=Xinanjiang`   |
 | b_Xinanjiang_shape_parameter=1  | *double* |   |   | parameter_adjustable  | direct runoff | when `surface_partitioning_scheme=Xinanjiang`   |
 | x_Xinanjiang_shape_parameter=1  | *double* |   |   | parameter_adjustable | direct runoff | when `surface_partitioning_scheme=Xinanjiang`   |
-
+| aet_rootzone                    | *bolean* | True, true or 1  |  | coupling parameter | `rootzone-based AET` | when `CFE coupled to SoilMoistureProfile` |
```
Comment (rlmcdaniel) [inline README.md:65]: "boolean" instead of "bolean"

File: README.md [hunk 3]
```diff
@@ -44,79 +62,75 @@ A [configs/](./configs/) directory contains primiary configuration text files fo
...
+| sft_coupled                     | *bolean* | True, true or 1  |  | coupling parameter | `ice-fraction based runoff` | when `CFE coupled to SoilFreezeThaw`|
```
Comment (rlmcdaniel) [inline README.md:66]: "boolean" instead of "bolean"

File: README.md [hunk 4]
```diff
@@ -44,79 +62,75 @@ A [configs/](./configs/) directory contains primiary configuration text files fo
...
 
-1. `gcc -lm ./src/main_cfe_aorc_pet.c ./forcing_code/src/pet.c ./forcing_code/src/bmi_pet.c ./src/cfe.c ./src/bmi_cfe.c ./forcing_code/src/aorc.c ./forcing_code/src/bmi_aorc.c -o run_cfe_aorc_et_bmi`. This generates an executable called `run_cfe_aorc_et_bmi`.
+1. `gcc -lm ./src/main_cfe_aorc_pet.c ./forcing_code/src/pet.c ./forcing_code/src/bmi_pet.c ./src/cfe.c ./src/bmi_cfe.c ./src/giuh.c ./forcing_code/src/aorc.c ./forcing_code/src/bmi_aorc.c -o run_cfe_aorc_et_bmi`. This generates an executable called `run_cfe_aorc_et_bmi`.
 2. To run this executable you must pass the path to the corresponding configuration files for CFE, PET and AORC (in that order):  `./run_cfe_aorc_et_bmi ./configs/cat_89_bmi_config_cfe_pass.txt ./configs/cat_89_bmi_config_aorc.txt ./configs/cat_89_bmi_config_pet_pass.txt`
 
-## Direct Runoff 
-In CFE the user has the option to pick a particular direct runoff (aka surface partitioning) method:
+### 4. CFE rootzone-based example couples C and C++ modules and can't be build without cmake
```
Comment (rlmcdaniel) [inline README.md:88]: "built" instead of "build"

File: README.md [hunk 5]
```diff
@@ -44,79 +62,75 @@ A [configs/](./configs/) directory contains primiary configuration text files fo
...
+## Rootzone-based Actual Evapotranspiration (AET)
+The user has the option to turn ON and OFF rootzone-based AET, default option is OFF. To turn it ON, the following parameters need to be included in the configuration file.
+1. aet_rootzone=true
+2. soil_layer_depths=0.1,0.4,1.0,2.0 (this is just an example)
+3. max_root_zone_layer=2
+
+## Soil freeze-thaw model (SFT)
+The Soil Freeze-Thaw (SFT) model is a standalone model.  For detailed information please refer to the [SFT repo](https://github.com/NOAA-OWP/SoilFreezeThaw). A few things to note when coupling SFT to CFE:
+1. SFT model provides `ice fraction` to CFE runoff schemes (Schaake `ice_fraction_schaake` and Xinanjiang `ice_fraction_xinan`)
+2. If the runoff scheme is Xinanjiang, no additional parameters are needed in the CFE config files. The framework will pass the parameter `ice_fraction_xinan` to CFE.
```
Comment (rlmcdaniel) [inline README.md:113]: It is unclear to me how to turn on SFT with the Xinanjiang scheme.

**Comment by rlmcdaniel** [comment]
In the readme there seem to be some inconsistencies with how things are referred to either within the document or between the document and what is generated by cmake using your CMakeLists file. Though I think someone could make it through running and compiling without changes, I think some changes for consistency would make it more clear and easier for the user. See below for some examples:
1.  Option 1 in the Compiling and Running (C&R) CFE is "BASE" which runs it's own forcing file, but in the Alternative C&R CFE section, it is referred to as "Read local forcing file" - I would refer to these the same way so it is clear you should get the same results and define it the same (i.e. with the same wording).  I would do the same with the others -- again so it is clear they are the same options and no assumptions need to be made.
2.   your cmake creates the run files in the build folder, so line 8 of C&R CFE where you navigate back to the cfe main folder with `cd ..` isn't needed.  Also, the run file name is different than in the readme and the path to the config files needs to be specified which isn't stated.  I think this can be figured out using the current doc, but can be made more clear
3.   the names for the run files in the alternative C&R CFE are different.  Not very important, but might be nice if they were the same as those generated by cmake for consistency; however, since the user can input whatever they want there, it isn't super important.

File: src/main_cfe_aorc_pet_rzAET.cxx [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    
+
+  }
+
+  // Run the Mass Balance check
+  mass_balance_check(cfe);
+  /************************************************************************
+    Finalize both the CFE and AORC bmi models
+  ************************************************************************/
+  printf("Finalizing BFE and AORC models\n");
```
Comment (rlmcdaniel) [inline src/main_cfe_aorc_pet_rzAET.cxx:287]: "CFE" instead of "BFE"

**Comment by rlmcdaniel** [comment]
Cloned and checked BASE, FORCING, FORCINGPET, and AETROOTZONE options in standalone.  All four options are working in standalone on UCS2.

**Comment by rlmcdaniel** [comment]
Not enough direction on connecting SFT.  I may have passed over it, but perhaps just direct folks to the SFT repository since it seems to have directions.

**Comment by rlmcdaniel** [comment]
Tested SFT with Schaake.  Worked and passed mass balance.

**Comment by rlmcdaniel** [comment]
ngen (cfe-modifications) with cfe (sft_aet_giuh_merge) using
data/example_realization_config_w_bmi_c__linux.json
appears to match ngen (master) with cfe (master)

File: README.md (hunk 1 above)
Comment (ajkhattak) [reply README.md:17]: done

File: README.md (hunk 2 above)
Comment (ajkhattak) [reply README.md:65]: done

File: README.md (hunk 3 above)
Comment (ajkhattak) [reply README.md:66]: done

File: README.md (hunk 4 above)
Comment (ajkhattak) [reply README.md:88]: done

File: src/main_cfe_aorc_pet_rzAET.cxx (hunk 6 above)
Comment (ajkhattak) [reply src/main_cfe_aorc_pet_rzAET.cxx:287]: this file has been deleted (not needed).

File: README.md (hunk 5 above)
Comment (ajkhattak) [reply README.md:113]: SFT related instructions are provided on SFT repo, and there is a parameter called `ice_fraction_scheme` for it.

**Comment by rlmcdaniel** [comment]
All PR examples match master examples in ngen.

File: README.md [hunk 7]
```diff
@@ -1,18 +1,46 @@
...
 
-## Compiling and running CFE
+## Compiling and Running CFE
+There are four examples for running CFE as described below. They assume you have [GCC](https://gcc.gnu.org) and [CMAKE](https://cmake.org/) on your machine. To build without cmake see section `Alternative: Compiling and Running CFE` below.
+
+1. `Option BASE` : CFE reads local forcing data (standalone CFE BMI run; one BMI example)
+2. `Option FORCING` : CFE uses an external module to read in a forcing file and pass those data using BMI (two BMIs example)
+3. `Option FORCINGPET` : CFE uses external modules to read in a forcing file and calculate potential evapotranspiration, and pass those data using BMIs (three BMIs example)
+4. `Option AETROOTZONE` : Option #3 running with rootzone-based actual evapotranspiration (AET). This version of CFE is coupled to the [SoilMoistureProfiles](https://github.com/NOAA-OWP/SoilMoistureProfiles) module which provides soil moisture (1D array) that is used to estimate actual evapotranspiration (AET) from the deepest rootzone layer. This example requires the user to clone [SoilMoistureProfiles](https://github.com/NOAA-OWP/SoilMoistureProfiles) repo. (four BMIs example)
```
Comment (rlmcdaniel) [inline README.md:17]: (four BMI examples) instead of (four BMIs example)

File: README.md [hunk 8]
```diff
@@ -44,79 +72,76 @@ A [configs/](./configs/) directory contains primiary configuration text files fo
...
+If the **Xinanjiang** scheme is choosen, three parameters need to be included in the configuration file:
+1. a_Xinanjiang_inflection_point_parameter
+2. b_Xinanjiang_shape_parameter
+3. x_Xinanjiang_shape_parameter 
+
+## Rootzone-based Actual Evapotranspiration (AET)
+The user has the option to turn ON and OFF rootzone-based AET, default option is OFF. To turn it ON, the following parameters need to be included in the configuration file.
+1. aet_rootzone=true
+2. soil_layer_depths=0.1,0.4,1.0,2.0 (this is just an example)
+3. max_root_zone_layer=2
```
Comment (rlmcdaniel) [inline README.md:118]: This is also just an example

File: README.md (hunk 7 above)
Comment (ajkhattak) [reply README.md:17]: I have reworded it to "example using four BMIs"

File: src/giuh.c [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  runoff_queue_m_per_timestep[N]=0.0;
+  
+  for(i=0;i<N;i++)
+    {
+      runoff_queue_m_per_timestep[i]+=giuh_ordinates[i]*runoff_m;
+    }
+  
+  runoff_m_now=runoff_queue_m_per_timestep[0];
+  
+  for(i=1;i<N;i++)  // shift all the entries in preperation ffor the next timestep
```
Comment (rlmcdaniel) [inline src/giuh.c:31]: remove extra "f" in ffor

File: src/bmi_cfe.c [hunk 10]
```diff
@@ -850,12 +961,99 @@ int read_init_config_cfe(const char* config_file, cfe_state_struct* model, doubl
...
+      for (int i=1; i <= model->soil_reservoir.n_soil_layers; i++){
+        current_depth = model->soil_reservoir.soil_layer_depths_m[i];
+        if (current_depth <= previous_depth)
+	  printf("WARNING: soil depths may be out of order.  One or more soil layer depths is less than or equal to the previous layer. Check CFE config file.\n");
+        model->soil_reservoir.delta_soil_layer_depth_m[i] = current_depth - previous_depth;
+        previous_depth = current_depth;
+      }
+      // Solve for the soil water content at field capacity via Clapp-Hornberger. See "Parameter Estimation for a Conceptual
+      // Functional Equivalent (CFE) Formulation of the National Water Model" equations 1-3 for detailed description. 
+													   double base = (*alpha_fc * STANDARD_ATMOSPHERIC_PRESSURE_PASCALS)/(WATER_SPECIFIC_WEIGHT * model->NWM_soil_params.satpsi);
```
Comment (rlmcdaniel) [inline src/bmi_cfe.c:1044]: Extra spaces not needed

Comment (ajkhattak) [reply src/bmi_cfe.c:1044]: done

File: src/giuh.c (hunk 9 above)
Comment (ajkhattak) [reply src/giuh.c:31]: done

File: README.md (hunk 8 above)
Comment (ajkhattak) [reply README.md]: modified this part

**Review by rlmcdaniel — APPROVED** [review]
All standalone and ngen tests passed, six tests total.

**Pull request merged**